File: codec/python/dcg_codec/codec/decode.py
import io
from codec import PREFIX, VERSION, LANGUAGE, compute_checksum, base36_to_char
from base64 import urlsafe_b64decode


def _get_u8(deck_bytes):
    return deck_bytes.read1(1)[0]


def _get_u32(deck_bytes):
    return deck_bytes.read1(4)


def _read_bits_from_byte(current_byte, mask_bits, delta_shift, out_bits):
    return ((current_byte & ((1 << mask_bits) - 1)) << delta_shift) | out_bits


def _is_carry_bit(current_byte, mask_bits):
    return 0 != current_byte & (1 << mask_bits)


def _read_encoded_u32(base_value, current_byte, delta_shift, deck_bytes):
    if delta_shift - 1 == 0 or _is_carry_bit(current_byte, delta_shift - 1):
        while True:
            next_byte = _get_u8(deck_bytes)
            base_value = _read_bits_from_byte(
                next_byte, 8 - 1, delta_shift - 1, base_value
            )
            if not _is_carry_bit(next_byte, 8 - 1):
                break
            delta_shift += 8 - 1
    else:
        return _read_bits_from_byte(current_byte, delta_shift - 1, 0, 0)
    return base_value


def _deserialize_card(
    version, deck_bytes, card_set, card_set_padding, prev_card_number
):
    current_byte = _get_u8(deck_bytes)
    card_count = (current_byte >> 6) + 1 if version == 0 else current_byte + 1
    current_byte = current_byte if version == 0 else _get_u8(deck_bytes)
    card_parallel_id = current_byte >> 3 & 0x07 if version == 0 else current_byte >> 5
    delta_shift = 3 if version == 0 else 5
    card_number = _read_bits_from_byte(current_byte, delta_shift - 1, 0, 0)
    prev_card_number += _read_encoded_u32(
        card_number, current_byte, delta_shift, deck_bytes
    )
    card = {
        "number": f"{card_set}-{str(prev_card_number).zfill(card_set_padding)}",
        "count": card_count,
    }
    if card_parallel_id != 0:
        card |= {"parallel-id": card_parallel_id}
    return (prev_card_number, card)


def _parse_deck(deck_bytes):
    version_and_digi_egg_count = _get_u8(deck_bytes)
    version = version_and_digi_egg_count >> 4
    assert VERSION >= version, f"Deck version {version} not supported"
    digi_egg_set_count = version_and_digi_egg_count & (
        0x07 if (3 <= version and version <= 4) else 0x0F
    )
    checksum = _get_u8(deck_bytes)
    deck_name_length = _get_u8(deck_bytes)
    language_number = (
        deck_name_length >> 6
        if version >= 5
        else version_and_digi_egg_count >> 3 & 0x01
    )
    if version >= 5:
        deck_name_length &= 0x3F
    total_card_bytes = (
        len(deck_bytes.getbuffer()[deck_bytes.tell() :]) - deck_name_length
    )

    language = LANGUAGE.get(language_number, "en")

    header_size = deck_bytes.tell()

    computed_checksum = compute_checksum(
        total_card_bytes,
        deck_bytes.getbuffer()[header_size : total_card_bytes + header_size],
    )
    assert (
        checksum == computed_checksum
    ), f"Deck checksum failed. {checksum} != {computed_checksum}"

    sideboard_byte = _get_u8(deck_bytes) if version >= 2 else 0
    has_icon = version >= 4 and (sideboard_byte >> 7) > 0
    sideboard_count = sideboard_byte & 0x7F if version >= 4 else sideboard_byte

    cards = []

    while len(deck_bytes.getbuffer()[deck_bytes.tell() :]) > deck_name_length:
        # Card Set Header
        # - Card Set
        if version == 0:
            card_set = str(_get_u32(deck_bytes).decode("UTF-8")).strip()
        else:
            card_set = ""
            current_byte = _get_u8(deck_bytes)
            card_set += base36_to_char(current_byte & 0x3F)
            while current_byte >> 7 != 0:
                current_byte = _get_u8(deck_bytes)
                card_set += base36_to_char(current_byte & 0x3F)
        # - Card Set Zero Padding and Count
        padding_and_set_count = _get_u8(deck_bytes)
        card_set_padding = (padding_and_set_count >> 6) + 1
        card_set_count = (
            _read_encoded_u32(
                padding_and_set_count, padding_and_set_count, 6, deck_bytes
            )
            if version >= 2
            else padding_and_set_count & 0x3F
        )

        prev_card_number = 0
        for _ in range(card_set_count):
            prev_card_number, card = _deserialize_card(
                version, deck_bytes, card_set, card_set_padding, prev_card_number
            )
            cards.append(card)

    icon = None
    deck_name = (
        bytes(deck_bytes.getbuffer()[deck_bytes.tell() :]).decode("UTF-8").strip()
    )
    deck = {
        "digi-eggs": cards[:digi_egg_set_count],
        "deck": cards[digi_egg_set_count : (len(cards) - sideboard_count)],
        "name": deck_name,
    }
    if version >= 3:
        deck |= {"language": language}
    if has_icon:
        icon, deck_name = deck_name[:8], deck_name[8:].strip()
        deck |= {"icon": icon, "name": deck_name}
    has_sideboard = version >= 2 and sideboard_count != 0
    if has_sideboard:
        deck |= {"sideboard": cards[(len(cards) - sideboard_count) :]}
    return deck


def decode(s):
    """
    Decode public function that takes a deck code and decodes to a Deck dict
    """
    prefix, deck_code = s[: len(PREFIX)], s[len(PREFIX) :]
    assert prefix == PREFIX, "Prefix was not 'DCG'"
    deck_code_bytes = deck_code.encode("UTF-8")
    deck_bytes = urlsafe_b64decode(
        (deck_code_bytes + (b"=" * (4 - (len(deck_code_bytes) % 4))))
    )
    return _parse_deck(io.BytesIO(deck_bytes))

File: codec/ruby/lib/dcg_codec/decode.rb
# frozen_string_literal: true

require 'base64'
require_relative 'common'

module DCGCodec
  # DCG Codec Decoder to handle decoding deck code strings
  class Decoder
    def initialize(deck_code)
      @deck_code = deck_code
      @offset = 0
    end

    def decode
      unless @deck_code.start_with?(DCGCodec::Common::PREFIX)
        raise "DECK_CODE must start with #{DCGCodec::Common::PREFIX}"
      end

      @deck_code.slice!(0, DCGCodec::Common::PREFIX.length)
      @deck_bytes = IO::Buffer.for(Base64.urlsafe_decode64(@deck_code))

      parse_deck
    end

    private

    def read_u8
      v = @deck_bytes.get_value(:U8, @offset)
      @offset += 1
      v
    end

    def read_4_u8
      v = @deck_bytes.get_values(%i[U8 U8 U8 U8], @offset)
      @offset += 4
      v
    end

    def read_bits_from_byte(current_byte, mask_bits, delta_shift, out_bits)
      ((current_byte & ((1 << mask_bits) - 1)) << delta_shift) | out_bits
    end

    def carry_bit?(current_byte, mask_bits)
      !(current_byte & (1 << mask_bits)).zero?
    end

    def read_encoded_u32(base_value, current_byte, delta_shift)
      if (delta_shift - 1).zero? || carry_bit?(current_byte, delta_shift - 1)
        loop do
          next_byte = read_u8
          base_value = read_bits_from_byte(
            next_byte, 8 - 1, delta_shift - 1, base_value
          )
          break unless carry_bit?(next_byte, 8 - 1)

          delta_shift += 8 - 1
        end
      end
      unless (delta_shift - 1).zero? || carry_bit?(current_byte, delta_shift - 1)
        return read_bits_from_byte(current_byte, delta_shift - 1, 0, 0)
      end

      base_value
    end

    def deserialize_card(version, card_set, card_set_padding, prev_card_number)
      current_byte = read_u8
      card_count = version.zero? ? (current_byte >> 6) + 1 : current_byte + 1
      current_byte = version.zero? ? current_byte : read_u8
      card_parallel_id = version.zero? ? current_byte >> 3 & 0x07 : current_byte >> 5
      delta_shift = version.zero? ? 3 : 5
      card_number = read_bits_from_byte(current_byte, delta_shift - 1, 0, 0)
      prev_card_number += read_encoded_u32(card_number, current_byte, delta_shift)
      card = {
        'number' => "#{card_set}-#{prev_card_number.to_s.rjust(card_set_padding, '0')}",
        'count' => card_count
      }
      card.store('parallel-id', card_parallel_id) unless card_parallel_id.zero?
      [prev_card_number, card]
    end

    def parse_deck
      version_and_digi_egg_count = read_u8
      version = version_and_digi_egg_count >> 4
      raise "Deck version #{version} not supported" unless version <= DCGCodec::Common::VERSION

      digi_egg_set_count = version_and_digi_egg_count & (version >= 3 && version <= 4 ? 0x07 : 0x0F)
      checksum = read_u8
      deck_name_length = read_u8
      language_number = version >= 5 ? deck_name_length >> 6 : version_and_digi_egg_count >> 3 & 0x01
      deck_name_length &= 0x3F if version >= 5
      total_card_bytes = @deck_bytes.size - @offset - deck_name_length
      language = DCGCodec::Common::LANGUAGE.fetch(language_number, 'en')
      header_size = @offset

      computed_checksum = DCGCodec::Common.compute_checksum(
        @deck_bytes.values(:U8, header_size, total_card_bytes)
      )
      raise "Deck checksum failed. #{checksum} != #{computed_checksum}" unless checksum == computed_checksum

      sideboard_byte = version >= 2 ? read_u8 : 0
      has_icon = version >= 4 && (sideboard_byte >> 7).positive?
      sideboard_count = version >= 4 ? sideboard_byte & 0x7F : sideboard_byte

      cards = []

      loop do
        # Card Set Header
        # - Card Set
        card_set = ''
        if version.zero?
          card_set = read_4_u8.pack('C*')
        else
          card_set = card_set.dup
          loop do
            current_byte = read_u8
            card_set << DCGCodec::Common.base36_to_char(current_byte & 0x3F)
            break if (current_byte >> 7).zero?
          end
        end
        # - Card Set Zero Padding and Count
        padding_and_set_count = read_u8
        card_set_padding = (padding_and_set_count >> 6) + 1
        card_set_count = if version >= 2
                           read_encoded_u32(padding_and_set_count, padding_and_set_count, 6)
                         else
                           padding_and_set_count & 0x3F
                         end

        prev_card_number = 0
        (1..card_set_count).each do |_|
          prev_card_number, card = deserialize_card(version, card_set, card_set_padding, prev_card_number)
          cards.push(card)
        end

        break if @offset >= @deck_bytes.size - deck_name_length
      end

      deck_name = @deck_bytes.get_string(@offset, deck_name_length)
      deck = {
        'digi-eggs' => cards.take(digi_egg_set_count),
        'deck' => cards[digi_egg_set_count, cards.length - digi_egg_set_count - sideboard_count],
        'name' => deck_name
      }
      deck.store('language', language) if version >= 3
      if has_icon
        icon = deck_name[0, 8].strip
        deck_name = deck_name[8, deck_name.length - 8].strip
        deck.merge({ 'icon' => icon, 'name' => deck_name })
      end
      has_sideboard = version >= 2 && !sideboard_count.zero?
      deck.store('sideboard', cards[-sideboard_count, sideboard_count]) if has_sideboard
      deck
    end
  end
end
